src/hotspot/cpu/x86/macroAssembler_x86.cpp

```diff
@@ -8601,16 +8605,22 @@ void MacroAssembler::char_array_compress(Register src, Register dst, Register le
     evmovdquw(tmp1Reg, mask2, Address(src, 0), /*merge*/ false, Assembler::AVX_512bit);
     evpcmpw(mask1, mask2, tmp1Reg, tmp2Reg, Assembler::le, /*signed*/ false, Assembler::AVX_512bit);
     ktestd(mask1, mask2);
-    jcc(Assembler::carryClear, return_zero);
+    jcc(Assembler::carryClear, copy_tail);
 
     evpmovwb(Address(dst, 0), mask2, tmp1Reg, Assembler::AVX_512bit);
     jmp(return_length);
 
+    bind(reset_for_copy_tail);
+    lea(src, Address(src, tmp5, Address::times_2));
+    lea(dst, Address(dst, tmp5, Address::times_1));
+    subptr(len, tmp5);
+    jmpb(copy_chars_loop);
```

Comment:
This cause a crash if I run with `-XX:UseAVX=3 -XX:AVX3Threshold=0`:

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (macroAssembler_x86.hpp:122), pid=3400452, tid=3400470
#  guarantee(this->is8bit(imm8)) failed: Short forward jump exceeds 8-bit offset at <null>:0
#
```

Needs to be a `jmp(copy_chars_loop)`.

Reply:
Alternatively:
```
if (UseSSE42Intrinsics) {
  jmpb(copy_chars_loop);
} else {
  jmp(copy_chars_loop);
}
```

More generally I do wonder if it'd make most sense to make the AVX512 and SSE42 implementations exclusive, though. Especially since we shouldn't mix AVX and SSE code (the code in this intrinsic seem to follow paths which are either/or, but it seems fragile). Perhaps @TobiHartmann can advise?

Reply:
> This cause a crash if I run with -XX:UseAVX=3 -XX:AVX3Threshold=0:

Good catch! Do we have a test for that scenario? If not, one should be added.

> Alternatively [...]

I would suggest to use `jmp(copy_chars_loop)` here for consistency with surrounding code.

> More generally I do wonder if it'd make most sense to make the AVX512 and SSE42 implementations exclusive

We don't mix AVX and SSE code here, right? We just fall back to SSE42 when AVX512 is not available or when the remaining length is below a threshold. Are you suggesting to better encapsulate both implementations (for example, factor them out into separate methods) or only ever using one?

Reply:
No, we don't mix: the SSE code is used as fallback only when the length is below 32 (if length is above 32 we check the tail with AVX code by shifting). 

I would suggest factoring out so that the implementations don't mix as much, mainly to reduce the number of possible variants to test and not to constrain one too much with the design of the other.  We now have AVX3-only, AVX3+SSE, SSE-only and plain, and I suggest dropping AVX3+SSE and fixing the AVX3-only so that it more efficiently handles strings of length 16-31 by duplicating (or using AVX instructions for copying 16 and 8 chars at once. Some code duplication perhaps, but simpler flow through each variant.

Reply:
That seems reasonable but would be out of scope for this RFE.

Reply:
Yes, for now we mainly need to make sure this works. There are a few regressions in microbenchmarks that I'm trying to get on top of, and the x64 intrinsics in particular seem problematic, but it seems reasonable to not hold up this PR and work on such improvements separately.